### ISSUE TITLE
Software Stack für HTTP Backend

### ISSUE THREAD

**Issue by schlamar** [WLANThermo/WLANThermo_v3#13]
Hallo zusammen,

ist der aktuelle Software Stack für das HTTP Backend schon in Stein gemeißelt? Ich habe sehr gute Erfahrungen mit der Kombination Python 3 + asyncio + aiohttp gemacht (auch auf dem Pi Zero). Großer Vorteil wäre hier, dass die ganze Abfragen parallel verarbeitet werden können. Das Frontend schickt ja die ganzen Anfragen ebenfalls parallel und sollte dann deutlich performanter laufen.

Auch weitere Sachen wie z.B. die Push-Nachrichten kann man hier sehr schön asynchron abhandeln und blockiert nicht mehr die Prozesse.

Gruß Marc

**Comment by BjoernSch** [comment]
Ja, aber ich habe es nicht geschafft da Threads im Hintergrund vernünftig zu integrieren, weil ich die nicht in asyncio realisiert bekommen habe.
Außerdem funktioniert sqlalchemy ORM auch nicht mit asyncio.
Hab's dann aufgegeben und für die minimale Anzahl an parallelen Sessions reicht auch Flask. Die Idee wäre auch den integrierten Webserver zu nutzen. Machen andere auch so und scheint auch für etwas anspruchsvollere Projekte zu funktionieren.
Pushnachrichten etc. würde ich per Threads in den Hintergrund verlegen, getriggert per Eventsystem.
Ich glaube das sollte ganz gut funktionieren und man schreibt nicht die Hälfte doppelt in synchron/ asynchron. Da lief es bei mir zumindest dann irgendwann drauf hinaus.
Für Standardapps fand ich aiohttp aber ziemlich gut, und ich dachte auch das es gut funktionieren könnte. Aber jede Schnittstelle zwischen asynchron und synchron hat dabei zu Kopfschmerzen geführt.
Es wäre aber gut noch einen Mitstreiter zu haben, auch wenn es "nur" Flask wird. Würde mich wirklich freuen.

**Comment by schlamar** [comment]
Synchrone Sachen kann man eigentlich problemlos in einen Thread-Pool auslagern:

    yield from loop.run_in_executor(None, function, *args)

Bzw. mit async / await und Python 3.5 noch schöner:

    await loop.run_in_executor(None, function, *args)

Das oben genannte sollte auch problemlos mit SQLAlchemy funktionieren, da hier ebenfalls ein Threadpool verwendet wird, wenn es die Datenbank unterstützt.

Der große Vorteil ist ja eigentlich, dass man auch für synchrone Sachen Threads nicht mehr explizit implementieren muss, was den Code deutlich vereinfacht und übersichtlicher macht. 

Ich habe Zuhause noch ein Real World Beispiel (Gärbox Steuerung über Pi Zero mit Webserver und o.g. Software Stack), das kann ich dir mal noch zukommen lassen. Vielleicht lässt du dich ja noch überzeugen :) 

Im ersten Schritt kann man aber auch erst mal alles sehr modular aufbauen, so dass der Austausch vom HTTP Backend dann auch mit relativ wenig Aufwand (z.B. in einer Version 3.1) erledigt werden könnte. D.h. keine Logik in der Flask Applikation, sondern alles ausgelagert in separate Funktionen/Klassen, die dann auch von einem aiohttp Backend genutzt werden könnten.

An Mitarbeit bin ich auf jeden Fall interessiert. Allerdings hab ich nur einen Thermo, der produktiv genutzt wird. Aber mit einer Debian VM sollte sich ja auch einiges entwickeln lassen.

**Comment by schlamar** [comment]
Hier findest du das versprochene Beispiel: https://gist.github.com/schlamar/a10fae7e6c731981827d6ab031f9fede 

Kommunikation Browser - Server ist über einen Websocket. Das hat den Vorteil, dass man nicht ständig pollen muss, sondern der Server direkt Updates in den Browser pushen kann.

**Comment by BjoernSch** [comment]
OK. So weit war ich auch schon.
Für Websockets bräuchten wir synchron auch irgendwelche Hilfsmittel, das hätte dann hier Vorteile.
Im Prinzip ist es also vermutlich egal, auch wenn man zu asyncio noch nicht so viele gute Beispiele findet.
Was mich noch stört:
- Schnittstellen zwischen Threads/ Coroutinen. Es wäre wie gesagt super wenn alles in einen Prozess passen würde. Das funktioniert aber nicht alles nur mit Coroutinen, ein paar Threads wird es wohl geben.
Auch hier wieder Vorteile/ Nachteile.
Was ich noch nicht gelöst habe:
Wie man sauber Exceptions von im Hintergrund laufenden Coroutinen einsammelt und sauber darauf reagiert.
